### PR TITLE
TPLink Archer c5v4 with new RTL8367C/S Giga Switch

### PR DESCRIPTION
**Pull request by ggbruno**
This is a PR to include RTL8367C/S giga switch and TPLink Archer C5 V4 Router that uses this switch.
It is just a rebase of the adaptation made by @981213 in PR #2174.
https://git.openwrt.org/?p=openwrt/staging/981213.git;a=shortlog;h=refs/heads/tp-ac5-v4

Everything works as expected.

Tested on:
TPLink Archer C5 V4 (MT7620 with RTL8367C/S)
TPLink Archer C2 V1 (MT7620 with RTL8367B/R)

### CODE REVIEW

**Comment by ggbruno** [comment]
Including @luizluca from #4010
Including @arinc9 from #4178

Those PR are related to this one.

**Comment by ggbruno** [comment]
Calling @dengqf6 (from original PR)

**Comment by arinc9** [comment]
Now this is interesting. The existing driver for rtl8367s, under mediatek architecture, uses a different driver under the directory of "rtl8365c" and had "glue" code: `rtl8367s.c rtl8367s_dbg.c rtl8367s_mdio.c` to support rtl8367s. This PR instead modifies rtl8367b driver to add support for rtl8367s & rtl8367c.

@dengqf6, care to elaborate?

**Comment by ggbruno** [comment]
> Now this is interesting. The existing driver for rtl8367s, under mediatek architecture, uses a different driver under the directory of "rtl8365c" and had "glue" code: `rtl8367s.c rtl8367s_dbg.c rtl8367s_mdio.c` to support rtl8367s. This PR instead modifies rtl8367b driver to add support for rtl8367c.

Actually, rtl8367c and s are the same... Realtek changed the naming convention in 8367 ICs. The old IC is called RTL8367R/B. This new IC is called RTL8367S/C ...

**Comment by adschm** [comment]
So, why do we need another PR for the same thing now? Do you really think that after nobody was interested in reviewing the existing one, increasing the amount of stuff to review will help here?

**Comment by arinc9** [comment]
> So, why do we need another PR for the same thing now? Do you really think that after nobody was interested in reviewing the existing one, increasing the amount of stuff to review will help here?

This PR adds support for rtl8367s/c by modifying the rtl8367b driver on the trunk. My and @luizluca's PRs use a different implementation which would make ~100K lines of change on the codebase if merged. Therefore, this PR has the minimal affect compared to the PRs of mine and @luizluca's. It also adds support for TP-Link Archer C5 V4 device.

Since this achieves what I wanted in much simple manner, I'm closing my PR and will refer to this one instead.

**Comment by ggbruno** [comment]
> So, why do we need another PR for the same thing now? Do you really think that after nobody was interested in reviewing the existing one, increasing the amount of stuff to review will help here?

This PR is an update from @981213 openwrt staging tree. The other PR uses a driver for RTL8367S that is "not interesting" for reviewing. As @981213 said in PR #2174, the full driver is "a bit scary", so they adapt it to the existing rtl8367b. @dengqf6  closed the PR because he do not have a router with the switch chip anymore, and after that, no one have checked @981213 staging tree for C5v4.

https://github.com/openwrt/openwrt/pull/2174#issuecomment-700648612
https://github.com/openwrt/openwrt/pull/2174#issuecomment-787222860

As @dengqf6 had closed his PR and @981213 have asked if someone can test his staging, I rebased and opened the original PR for review....

File: target/linux/generic/files/drivers/net/phy/rtl8367b.c [hunk 1]
```diff
@@ -1021,6 +1051,16 @@ static int rtl8367b_setup(struct rtl8366_smi *smi)
 			RTL8367B_PORT_MISC_CFG_EGRESS_MODE_ORIGINAL <<
 				RTL8367B_PORT_MISC_CFG_EGRESS_MODE_SHIFT);
 
+	for (i = 0; i < RTL8367B_NUM_PHYS; i++) {
+		int data;
+		rtl8367b_read_phy_reg(smi, i, 0, &data);
+		/* release phy power down */
+		data &= ~BIT(11);
+		/* restart auto negotiation */
+		data |= BIT(9);
```
Comment (blocktrron) [inline target/linux/generic/files/drivers/net/phy/rtl8367b.c:1060]: Please use the macros from `linux/mii.h`

`BMCR_PDOWN` / `BMCR_ANRESTART`

File: target/linux/generic/files/drivers/net/phy/rtl8367b.c [hunk 2]
```diff
@@ -1021,6 +1051,16 @@ static int rtl8367b_setup(struct rtl8366_smi *smi)
 			RTL8367B_PORT_MISC_CFG_EGRESS_MODE_ORIGINAL <<
 				RTL8367B_PORT_MISC_CFG_EGRESS_MODE_SHIFT);
 
+	for (i = 0; i < RTL8367B_NUM_PHYS; i++) {
+		int data;
+		rtl8367b_read_phy_reg(smi, i, 0, &data);
```
Comment (blocktrron) [inline target/linux/generic/files/drivers/net/phy/rtl8367b.c:1056]: And while at it, you can demystify the register by using `MII_BMCR`.

**Comment by luizluca** [comment]
> > So, why do we need another PR for the same thing now? Do you really think that after nobody was interested in reviewing the existing one, increasing the amount of stuff to review will help here?
> 
> This PR adds support for rtl8367s/c by modifying the rtl8367b driver on the trunk. My and @luizluca's PRs use a different implementation which would make ~100K lines of change on the codebase if merged. Therefore, this PR has the minimal effect compared to the PRs of mine and @luizluca's. It also adds support for TP-Link Archer C5 V4 device.

The driver from closed @dengqf6 PR seems to be the best alternative as it shares more code with existing driver. All looks great but it didn't get merged.

My PR used the same driver unofficial images are using for years and mediatek target already uses in OpenWrt. Although it touches 100k lines, it is simply a file move from mediatek to generic. The actual code changes are minimal.

I beg to differ on less side effects argument. Although I think this is the best alternative, it will touch an existing driver that is already in use. My PR didn't change a single line of kernel code for existing devices. It might be worse but still safer.

> Since this achieves what I wanted in much simple manner, I'm closing my PR and will refer to this one instead.

We have two working drivers, tested and in use by many users. This is way better than most new devices.
I'm good with any alternative (this is the 4th PR). I just wish one of them gets merged.

I can help with tests and debug if needed but I think both drivers are already overtested.

**Comment by ggbruno** [comment]
> The driver from closed @dengqf6 PR seems to be the best alternative as it shares more code with existing driver. All looks great but it didn't get merged.

Yes, I saw that... @981213 have worked on some problems with the driver... As it modify an existing driver in OpenWrt, it must comply with formatting and stuff... From what I saw, @981213 said it needs testing... He have formatted the driver, but don't have any router for testing....

> My PR used the same driver unofficial images are using for years and mediatek target already uses in OpenWrt. Although it touches 100k lines, it is simply a file move from mediatek to generic. The actual code changes are minimal.

I agree with @981213 that the official driver from Realtek (the rtl8367c) is a bit scary to understand. It's like a black box. I don't know why it get merged in Mediatek repository. Maybe to have a working MT7622... I think, if this driver get merged, we can work to remove the Realtek black box from Mediatek target and use this one. It is simpler to review and easier to port to new RTL giga variants. 

> I beg to differ on less side effects argument. Although I think this is the best alternative, it will touch an existing driver that is already in use. My PR didn't change a single line of kernel code for existing devices. It might be worse but still safer.

That was one of the problems in #2174. But @981213 have worked to isolate the driver with DTS directives. I have already testes on two RTL8367B routers with no problem... I can create a new driver for RTL8367C based on this one that is a simple file (as the OpenWrt RTL8367B), but I think that the OpenWrt project can gain in code reuse, particularly if the driver is generic... And porting OpenWrt RTL8367* driver to DSA is easier...

**Comment by adschm** [comment]
Then please either close #4010 or this one. I do not see any gain from having two PRs rotting instead of one...

File: target/linux/ramips/dts/mt7620a_tplink_archer-c5-v4.dts [hunk 3]
```diff
@@ -0,0 +1,204 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/dts-v1/;
```
Comment (adschm) [inline target/linux/ramips/dts/mt7620a_tplink_archer-c5-v4.dts:2]: dts-v1 should be dropped.

File: target/linux/ramips/dts/mt7620a_tplink_archer-c5-v4.dts [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+	status = "okay";
+};
+
+&spi0 {
+	status = "okay";
+
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <60000000>;
```
Comment (adschm) [inline target/linux/ramips/dts/mt7620a_tplink_archer-c5-v4.dts:108]: > 60 MHz typically required m25p,fast-read ...?

File: target/linux/ramips/dts/mt7620a_tplink_archer-c5-v4.dts [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+&ehci {
+	status = "okay";
+};
+
+&ohci {
+	status = "okay";
+};
+
+&wmac {
+	ralink,mtd-eeprom = <&radio 0>;
```
Comment (adschm) [inline target/linux/ramips/dts/mt7620a_tplink_archer-c5-v4.dts:188]: 0x0 instead of "0"

File: target/linux/ramips/image/mt7620.mk [hunk 6]
```diff
@@ -1010,6 +1010,22 @@ define Device/tplink_archer-c2-v1
 endef
 TARGET_DEVICES += tplink_archer-c2-v1
 
+define Device/tplink_archer-c5-v4
+  $(Device/tplink-v2)
+  SOC := mt7620a
+  IMAGE_SIZE := 7808k
+  TPLINK_FLASHLAYOUT := 8Mmtk
+  TPLINK_HWID := 0x04DA857C
+  TPLINK_HWREV := 0x0C000600
```
Comment (adschm) [inline target/linux/ramips/image/mt7620.mk:1019]: Lower case is preferred for hex letters.

File: target/linux/ramips/image/mt7620.mk [hunk 7]
```diff
@@ -1010,6 +1010,22 @@ define Device/tplink_archer-c2-v1
...
+  TPLINK_HWREVADD := 0x04000000
+  IMAGES += tftp-recovery.bin
+  IMAGE/tftp-recovery.bin := pad-extra 128k | $$(IMAGE/factory.bin)
+  DEVICE_MODEL := Archer C5
+  DEVICE_VARIANT := v4
+  DEVICE_PACKAGES := kmod-usb2 kmod-usb-ohci kmod-usb-ledtrig-usbport kmod-mt76x2 kmod-switch-rtl8367b
```
Comment (adschm) [inline target/linux/ramips/image/mt7620.mk:1025]: Wrap after 74 chars. max

**Comment by arinc9** [comment]
> Then please either close #4010 or this one. I do not see any gain from having two PRs rotting instead of one...

Just weighing in my personal opinion: #4010 should be closed.

The realtek driver it includes requires mediatek architecture specific strings on DT file and only enables ramips & mediatek architectures to use the driver. It also uses the full “black box” driver.

**Comment by luizluca** [comment]
I always thought this approach (like @dengqf6 's) was technically better. The question is which one has better chance to get merged 😀. This is the 4th PR trying to merge a driver for that switch. I hope we don't get a 5th.

I'll close #4010. We just need any solution to get merged.

**Comment by arinc9** [comment]
I plan to add rtl8365mb support, which Asus RT-AC88U uses, with @ggbruno by modifying this driver. I hope this PR gets merged so we can add support for rtl8365mb (and RT-AC88U device support) on top of it.

File: target/linux/ramips/dts/mt7620a_tplink_archer-c5-v4.dts (hunk 4 above)
Comment (ggbruno) [reply target/linux/ramips/dts/mt7620a_tplink_archer-c5-v4.dts:107]: checked and fast-read is working. Changed as suggested.

**Comment by arinc9** [comment]
What's the hold up?

**Comment by arinc9** [comment]
@ggbruno Can you turn these commits into a patch and send it to openwrt-devel mailing list? It's very unlikely that this is going to be merged here.
Let me know if you need any help.

**Comment by luizluca** [comment]
> This may be a good solution... Maybe we get a better chance if first we port the switch to DSA, and then include the C5 port (using DSA)...

The issue with C5 is that it needs a new switch driver, that needs to be maintained. If we have an upstream driver, it would be great and nobody will be scared to merge the PR.

> Do you have access to Archer C5 V4 and Archer C2 V1? I can't get my hands on one of these anywhere. Let's try the DSA driver together.

I do have a C5. I can help with the tests. Is upstream RTL8365MB driver compatible with RTL8367C/S? Does it have already a proper package for OpenWrt?

**Comment by luizluca** [comment]
> I've put new images containing the mainline Linux kernel. Anyone with serial access should test it and let me know if the device boots fine. The switch won't work for now as we don't use swconfig anymore. https://github.com/arinc9/openwrt/releases/tag/master-realtek-work-tplink_archer-c5-v4

It is boot looping.

**Comment by arinc9** [comment]
> > I've put new images containing the mainline Linux kernel. Anyone with serial access should test it and let me know if the device boots fine. The switch won't work for now as we don't use swconfig anymore. https://github.com/arinc9/openwrt/releases/tag/master-realtek-work-tplink_archer-c5-v4
> 
> It is boot looping.

I've updated the images to run Linux 5.15-rc5. Let me know what happens with this one.

**Comment by nickbash11** [comment]
have a EC220-G5 with the similar hardware https://gist.github.com/nickbash11/67d3c63edff2906080f71bd971056eb8

**Comment by S7venLights** [comment]
> have a EC220-G5 with the similar hardware https://gist.github.com/nickbash11/67d3c63edff2906080f71bd971056eb8

Is it similar hardware or the same, I have an tp-link ec220-G5 ver:2.0 (It also says AC1200 on the box)
 is it the same hardware as [this](https://openwrt.org/toh/hwdata/tp-link/tp-link_archer_c5_ac1200_v2.0) or [this](https://openwrt.org/toh/hwdata/tp-link/tp-link_archer_c5_ac1200_v1)?
 
 As to this issue I'm commenting on, is Giga Switch to do with ethernet? Will Ethernet not work with OpenWRT currently?

**Comment by luizluca** [comment]
> > have a EC220-G5 with the similar hardware https://gist.github.com/nickbash11/67d3c63edff2906080f71bd971056eb8
> 
> Is it similar hardware or the same HW?
>  I have an tp-link ec220-G5 ver:2.0 (It also says AC1200 on the box)
>  is it the same hardware as [this](https://openwrt.org/toh/tp-link/archer_c5_v4)?
>  It does have the [same RAM and CPU Freq](https://openwrt.org/inbox/toh/tp-link/ec220-g5_v2)
>  
>  As to this issue I'm commenting on, is Giga Switch to do with ethernet? Will Ethernet not work with OpenWRT currently?
>  Seems like the EC220-G5 also has wifi speed issues, but is [okay in 5G at least](https://forum.openwrt.org/t/support-for-new-archer-c5-v4/15889/488)...

The support for c5v4 was not merged because there is no support for the internal switch. Without a switch, there is no Ethernet port. Normally a device with only wifi will not be merged as many procedures do require a wired connection, like downloading the packages to support 5g modems.

There is support for that switch in kernel upstream. It will eventually land in OpenWrt making the c5v4 support trivial to add. However, it will not support a bunch of hardware offload features. But it is better than nothing.

**Comment by arinc9** [comment]
Hey Luiz!

Were you ever able to work on implementing bridge offloading for the RTL8365MB DSA subdriver?

My folks recently received a EC220-G5 at the office. It may become my next device to play with. I especially want to see how the OpenWrt-maintained ethernet driver for MT7620 and the other OpenWrt-maintained driver for the 100M PHY MT7530 switch on the MT7620 SoC works.

**Comment by luizluca** [comment]
> Hey Luiz!
> 
> Were you ever able to work on implementing bridge offloading for the RTL8365MB DSA subdriver?

No, I wasn't in my plans. Alvin Šipraga mentioned he would work on that but  the last patches in kernel for that driver were in June. Maybe we can ping him.

The last time I tried rebasing my branch, Ethernet wasn't working. There are tons of DSA backported patches that might mess with each other. I was planning to play a little more with this device once kernel 6.x lands on OpenWrt.

> My folks recently received a EC220-G5 at the office. It may become my next device to play with. I especially want to see how the OpenWrt-maintained ethernet driver for MT7620 and the other OpenWrt-maintained driver for the MT7530 switch (with 100M PHYs) on the MT7620 SoC works.

Well, for C5v4, mt7620 just need to get out of the way. The MT7530 must be configured without VLAN filtering and possibly Mac learning (required in the case of RTL DSA without bridge offloading), making it act as a single Ethernet connected to the realtek switch. It does not use the 100M ports but a single internal 1GE. If you want to try the realtek DSA driver, all offloading features from mediatek must go as the RTL DSA tag breaks everything, but it is not different from the mt7621 situation that you know very well.

**Comment by arinc9** [comment]
> No, I wasn't in my plans. Alvin Šipraga mentioned he would work on that but the last patches in kernel for that driver were in June. Maybe we can ping him.

Can you CC me when you do?

File: target/linux/generic/files/drivers/net/phy/rtl8367b.c [hunk 8]
```diff
@@ -716,33 +734,46 @@ static int rtl8367b_write_phy_reg(struct rtl8366_smi *smi,
...
 
 	rlvid = (chip_ver >> RTL8367B_CHIP_VER_RLVID_SHIFT) &
 		RTL8367B_CHIP_VER_RLVID_MASK;
 
-	switch (rlvid) {
-	case 0:
-		initvals = rtl8367r_vb_initvals_0;
-		count = ARRAY_SIZE(rtl8367r_vb_initvals_0);
-		break;
+	if (smi->flags & RTL8367B_FLAG_MATCH_CHIP_NUM) {
```
Comment (andy9a9) [inline target/linux/generic/files/drivers/net/phy/rtl8367b.c:750]: `smi->flags` has undefined value, while it's assigned only inside `#ifdef CONFIG_OF`

File: target/linux/generic/files/drivers/net/phy/rtl8367b.c [hunk 9]
```diff
@@ -1540,14 +1581,26 @@ static int rtl8367b_detect(struct rtl8366_smi *smi)
...
 
-	switch (chip_ver) {
-	case 0x1000:
-		chip_name = "8367RB";
-		break;
-	case 0x1010:
-		chip_name = "8367R-VB";
-		break;
-	default:
+	if (smi->flags & RTL8367B_FLAG_MATCH_CHIP_NUM) {
```
Comment (andy9a9) [inline target/linux/generic/files/drivers/net/phy/rtl8367b.c:1584]: Same comment as above

**Review by namiltd — COMMENTED** [review]
Patch works seamlessly with v22.03.05

P.S. The mtd-mac-address parameter should be replaced with nvmem-cells
and mtd-mac-address-increment with mac-address-increment

File: target/linux/ramips/dts/mt7620a_tplink_archer-c5-v4.dts [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+
+&pcie0 {
+	wifi@0,0 {
+		reg = <0x0000 0 0 0 0>;
+		mediatek,mtd-eeprom = <&radio 0x8000>;
+		ieee80211-freq-limit = <5000000 6000000>;
+		mtd-mac-address = <&rom 0xf100>;
+		mtd-mac-address-increment = <2>;
+	};
+};
```
Comment (namiltd) [inline target/linux/ramips/dts/mt7620a_tplink_archer-c5-v4.dts:204]: mt7620a_tplink_archer-c5-v4.dts file compatible with versions 22.03.5 and 23.05.0:
```
// SPDX-License-Identifier: GPL-2.0-or-later OR MIT

#include "mt7620a_tplink_archer.dtsi"

/ {
	compatible = "tplink,archer-c5-v4", "ralink,mt7620a-soc";
	model = "TP-Link Archer C5 v4";

	aliases {
		led-boot = &led_power;
		led-failsafe = &led_power;
		led-running = &led_power;
		led-upgrade = &led_power;
		label-mac-device = &ethernet;
	};

	leds {
		compatible = "gpio-leds";

		led_power: power {
			label = "green:power";
			gpios = <&gpio2 2 GPIO_ACTIVE_LOW>;
		};

		wlan2g {
			label = "green:wlan2g";
			gpios = <&gpio0 14 GPIO_ACTIVE_LOW>;
			linux,default-trigger = "phy1tpt";
		};

		wlan5g {
			label = "green:wlan5g";
			gpios = <&gpio0 12 GPIO_ACTIVE_LOW>;
			linux,default-trigger = "phy0tpt";
		};

		wan {
			label = "green:wan";
			gpios = <&gpio0 10 GPIO_ACTIVE_LOW>;
		};

		wan_orange {
			label = "orange:wan";
			gpios = <&gpio0 9 GPIO_ACTIVE_LOW>;
		};

		lan {
			label = "green:lan";
			gpios = <&gpio0 8 GPIO_ACTIVE_HIGH>;
		};

		usb {
			label = "green:usb";
			gpios = <&gpio0 7 GPIO_ACTIVE_LOW>;
			trigger-sources = <&ohci_port1>, <&ehci_port1>;
			linux,default-trigger = "usbport";
		};

		wps {
			label = "green:wps";
			gpios = <&gpio0 1 GPIO_ACTIVE_LOW>;
		};
	};

	rtl8367s {
		compatible = "realtek,rtl8367c";
		cpu_port = <7>;
		realtek,extif2 = <1 0 1 1 1 1 1 1 2>;
		mii-bus = <&mdio0>;
		phy-id = <29>;
	};
};

&state_default {
	gpio {
		groups = "i2c", "uartf", "ephy", "rgmii2";
		function = "gpio";
	};
};

&ethernet {
	pinctrl-names = "default";
	pinctrl-0 = <&rgmii1_pins &mdio_pins>;

	nvmem-cells = <&macaddr_rom_f100>;
	nvmem-cell-names = "mac-address";

	port@5 {
		status = "okay";
		mediatek,fixed-link = <1000 1 1 1>;
		phy-mode = "rgmii";
	};

	mdio0: mdio-bus {
		status = "okay";
		reset-gpios = <&gpio2 20 GPIO_ACTIVE_LOW>;
		reset-delay-us = <10000>;
	};
};

&wmac {
	pinctrl-names = "default", "pa_gpio";
	pinctrl-0 = <&pa_pins>;
	pinctrl-1 = <&pa_gpio_pins>;

	ralink,mtd-eeprom = <&radio 0x0>;
	nvmem-cells = <&macaddr_rom_f100>;
	nvmem-cell-names = "mac-address";
};

&pcie0 {
	wifi@0,0 {
		reg = <0x0000 0 0 0 0>;
		mediatek,mtd-eeprom = <&radio 0x8000>;
		ieee80211-freq-limit = <5000000 6000000>;
		nvmem-cells = <&macaddr_rom_f100>;
		nvmem-cell-names = "mac-address";
		mtd-mac-address-increment = <2>;
	};
};

Comment (luizluca) [reply target/linux/ramips/dts/mt7620a_tplink_archer-c5-v4.dts:204]: Could you share these changes in a public git tree?

**Comment by namiltd** [comment]
Version compatible with release 23.05: https://github.com/openwrt/openwrt/pull/14314

**Comment by luizluca** [comment]
An alternative PR was just merged. #14418

This PR can rest in peace.

**Pull request closed without merging**